File: app/src/main/java/com/codemybrainsout/imageviewer/view/search/SearchPresenter.java
package com.codemybrainsout.imageviewer.view.search;

import com.codemybrainsout.imageviewer.api.SearchService;
import com.codemybrainsout.imageviewer.model.Photo;
import com.codemybrainsout.imageviewer.model.Search;
import com.codemybrainsout.imageviewer.model.User;

import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by ahulr on 11-06-2017.
 */

public class SearchPresenter implements SearchContract.Presenter {

    private SearchService searchService;
    private SearchContract.View view;

    public SearchPresenter(SearchService searchService, SearchContract.View view) {
        this.searchService = searchService;
        this.view = view;
    }

    @Override
    public void openUser(User user) {
        view.showUser(user);
    }

    @Override
    public void openPhoto(Photo photo) {
        view.showPhoto(photo);
    }

    @Override
    public void searchPhotos(String query, int page, int limit) {
        view.showLoading();
        searchService.searchPhoto(query, page, limit)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Search>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull Search search) {
                        List<Photo> photos = search.getResults();
                        view.refreshPhotos(photos);
                        view.hideLoading();
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        e.printStackTrace();
                        view.showError(e.getMessage());
                        view.hideLoading();
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void searchMorePhotos(String query, int page, int limit) {
        searchService.searchPhoto(query, page, limit)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Search>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull Search search) {
                        List<Photo> photos = search.getResults();
                        view.addPhotos(photos);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        e.printStackTrace();
                        view.showError(e.getMessage());
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/codemybrainsout/imageviewer/view/search/SearchPresenter.kt
package com.codemybrainsout.imageviewer.view.search

import com.codemybrainsout.imageviewer.api.SearchService
import com.codemybrainsout.imageviewer.model.Photo
import com.codemybrainsout.imageviewer.model.Search
import com.codemybrainsout.imageviewer.model.User
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ahulr on 11-06-2017.
 */
class SearchPresenter @Inject constructor(private val searchService: SearchService, val view: SearchContract.View) :
    SearchContract.Presenter {

    override fun openUser(user: User?) {
        view.showUser(user)
    }

    override fun openPhoto(photo: Photo?) {
        view.showPhoto(photo)
    }

    override fun searchPhotos(query: String?, page: Int, limit: Int) {
        view.showLoading()
        searchService.searchPhoto(query, page, limit)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                val photos: List<Photo>? = it.results
                photos?.let { photos -> view.refreshPhotos(photos) }
                view.hideLoading()
            }, {
                it.printStackTrace()
                view.showError(it.message)
                view.hideLoading()
            })
    }

    override fun searchMorePhotos(query: String?, page: Int, limit: Int) {
        searchService.searchPhoto(query, page, limit)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                val photos: List<Photo>? = it.results
                if (photos != null) {
                    view.addPhotos(photos)
                }
            }, {
                it.printStackTrace()
                view.showError(it.message)
            })
    }
}